router/internal/graphqlmetrics/batch_queue.go

```diff
@@ -0,0 +1,163 @@
+package graphqlmetrics
+
+import "time"
+
+// Inspired by https://github.com/wind-c/bqueue
+
+import (
+	"context"
+	"sync/atomic"
+)
+
+const (
+	defaultInterval      = time.Duration(5) * time.Second
+	defaultMaxBatchItems = 1024
+	defaultMaxQueueSize  = 1024
+)
+
+// BatchQueueOptions configure time interval and set a batch limit.
+type BatchQueueOptions struct {
+	Interval      time.Duration // wait time when batch quantity is insufficient
+	MaxBatchItems int           // maximum items per batch
+	MaxQueueSize  int           // maximum queue size
+}
+
+func (o *BatchQueueOptions) ensureDefaults() {
+	if o.Interval == 0 {
+		o.Interval = defaultInterval
+	}
+
+	if o.MaxBatchItems == 0 {
+		o.MaxQueueSize = defaultMaxBatchItems
+	}
+
+	if o.MaxQueueSize == 0 {
+		o.MaxQueueSize = defaultMaxQueueSize
+	}
+}
+
+// BatchQueue coordinates dispatching of queue items by time intervals
+// or immediately after the batching limit is met.
+type BatchQueue struct {
```

Comment:
How confident are you that the implementation is "good"?
I could try and analyze it more, but would like to first ask if you have evaluated existing solutions?
Is there no batch library that is tested and does what we need?

Reply:
I'll postpone evaluating the batch impl until answered.

Reply:
Most of the code and tests were copied from https://github.com/wind-c/bqueue. I researched alternatives, but this is the only implementation I found good enough for our use case. It has to support

- Non-blocking
- Dispatching by interval and batch limit

Reply:
Please test it :smiley_cat: